### ISSUE TITLE
Implement generic batch normalization layer.

### ISSUE THREAD

**Issue by veritas9872** [pytorch/pytorch#89116]
### 🚀 The feature, motivation and pitch

Currently, the SyncBatchNorm module in PyTorch only has a `convert_sync_batchnorm` method to convert other Batch Normalization layers to SyncBatchNorm. However, it does not implement any method to revert the layers to their original classes.

However, implementing this would require either saving the original class type inside the SyncBatchNorm or implementing 1d, 2d, and 3d implementations of SyncBatchNorm. It would be preferable to implement instead a generic BatchNorm layer that can handle arbitrary dimensions.

Therefore, I would like to propose a generic `BatchNorm` layer that can handle arbitrary dimensions and can convert any related BatchNorm layer to itself, similar to `convert_sync_batchnorm`. This would allow removing the SyncBatchNorm layer for non-distributed inference and can convert any previous models to the new BatchNorm layer. Since the current functional `batch_norm` function is dimension neutral, the implementation will not be complicated.

### Alternatives

Possibly save the original dimensions in the SyncBatchNorm layer. However, this would not work if the original model is unknown.

### Additional context

SyncBatchNorm causes issues on single-process CPU inference and is challenging to convert to ONNX, etc. Having a generic BatchNorm would be the minimally invasive approach. Other issues have also proposed this solution before.

https://github.com/pytorch/pytorch/issues/41081
https://github.com/pytorch/pytorch/issues/82464

**Comment by vadimkantorov** [comment]
Also related: https://github.com/pytorch/pytorch/issues/41243, https://github.com/pytorch/pytorch/issues/66073, https://github.com/pytorch/pytorch/issues/68648, https://github.com/pytorch/pytorch/issues/26288

**Comment by veritas9872** [comment]
@albanD @vadimkantorov  I have implemented the change, as well as most of the tests, but I do not have permission to push and create a new branch in PyTorch. Does this require a particular process?

**Comment by kshitij12345** [comment]
@veritas9872 That's great!

To make a PR, you can fork the repository and push the branch to your fork. Post that you can create a Pull Request from Github.

For other practices regarding linting your code you can refer: https://github.com/pytorch/pytorch/blob/master/CONTRIBUTING.md#contributing-to-pytorch

Thanks!

**Comment by veritas9872** [comment]
https://github.com/pytorch/pytorch/pull/89717#issuecomment-1328401719

**Comment by vadimkantorov** [comment]
My own proposal would be to unite SyncBatchNorm with _BatchNorm class (make sync = True a property/option + maybe a static method (which would simply set bn.sync = True), so that one can model.apply(BatchNorm.convert_to_sync) or similar) and also make the latter public (and still inherit from it BatchNorm2d and friends for compat)

**Comment by veritas9872** [comment]
@vadimkantorov I agree that uniting SyncBatchNorm with BatchNorm would be a better idea. Would this cause any issues with quantization, TorchScript conversion, etc.?

**Comment by vadimkantorov** [comment]
I'm not knowledgeable enough about this :( @albanD would know better!

A relevant issue search: https://github.com/pytorch/pytorch/issues?q=is%3Aissue+is%3Aopen+syncbatchnorm+

**Comment by yanbing-j** [comment]
We have a PoC of enabling SyncBatchNorm in CPU backend https://github.com/pytorch/pytorch/pull/78081. Will rebase it to get ready for review.

**Comment by rwightman** [comment]
> My own proposal would be to unite SyncBatchNorm with _BatchNorm class (make sync = True a property/option + maybe a static method (which would simply set bn.sync = True), so that one can model.apply(BatchNorm.convert_to_sync) or similar) and also make the latter public (and still inherit from it BatchNorm2d and friends for compat)

This would be really nice, a united batch norm, I have some use cases where it's desirable to inherit from BatchNorm and add some additional functionality, current `convert_sync_batchnorm` mechanism completely breaks that and requires a custom convert_sync_batchnorm fn. It'd be easy to work with a united approach.
